File: app/src/main/java/com/example/compass/ui/compass/CompassView.java
package com.example.compass.ui.compass;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import com.example.compass.R;

public class CompassView extends FrameLayout {

    private Integer currentCompassRotation = 0;

    private CompassNeedleView compassNeedleView;
    private CompassPlateView compassPlateView;

    public CompassView(Context context, AttributeSet attrs) {
        super(context, attrs);

        inflate(context, R.layout.compass_view, this);
        compassNeedleView = findViewById(R.id.compass_needle_view);
        compassPlateView = findViewById(R.id.compass_plate_view);
    }

    public void rotateCompass(Integer degree) {
        Integer rotationDelta = currentCompassRotation - degree;
        setRotation(-(currentCompassRotation + rotationDelta));
        currentCompassRotation = degree;
    }

    public void showAzimuthDirection(double azimuth) {
        compassNeedleView.rotateNeedle((float) azimuth);
        compassPlateView.drawLatLngIndicator(azimuth);
    }

}

File: app/src/main/java/com/example/compass/CompassView.kt
package com.example.compass

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.animation.Animation
import android.view.animation.RotateAnimation
import android.widget.LinearLayout
import kotlinx.android.synthetic.main.compass_view.view.*

class CompassView(context: Context,attrs:AttributeSet) : LinearLayout(context,attrs) {
    init {
        View.inflate(context, R.layout.compass_view, this)
    }

    var accelerometerValues : FloatArray? = null
    var magnetoMeterValues : FloatArray? = null

    var currentDegree = 0f

    fun rotateCompass(degree: Float) {

        val rotateAnimation = RotateAnimation(
            currentDegree,
            -degree,
            Animation.RELATIVE_TO_SELF,
            0.5f,
            Animation.RELATIVE_TO_SELF,
            0.5f
        ).apply {
            duration = 2000
            fillAfter = true
        }

        compass_image.startAnimation(rotateAnimation)

        currentDegree = -degree
    }

}